src/lib/PPrint.hs

```diff
@@ -39,21 +39,48 @@ import Name
 import Syntax
 import Parser (showPrimName)
 
--- Specifies what kinds of operations are allowed to be printed at this point.
--- Printing at AppPrec level means that applications can be printed
--- with no parentheses, but binops must be parenthesized.
-data PrecedenceLevel  = LowestPrec
-                      | AppPrec
-                      | ArgPrec  -- Only single symbols and parens allowed
-                      deriving (Eq, Ord)
-
+-- A DocPrec is a slightly context-aware Doc, specifically one that
+-- knows the precedence level of the immediately enclosing operation,
+-- and can decide to parenthesize itself accordingly.
+-- For example, when printing `x = f (g 1)`, we know that
+-- - We need parens around `(g 1)` because applying `f` binds
+--   tighter than applying `g` (because application is left-associative)
+-- - We do not need parens around `g` or 1, because there is nothing
+--   there that may bind less tightly than the function applications.
+-- - We also do not need parens around the whole RHS `f (g 1)`, because
+--   the `=` binds less tightly than applying `f`.
+--
+-- This is accomplished in the `Expr` instance of `PrettyPrec` by
+-- coding function application to require `ArgPrec` from the arguments
+-- (via the default behavior of `prettyFromPrettyPrec`), but to
+-- provide only `AppPrec` for the application expression itself.  The
+-- outer application is not wrapped in parens because the let binding
+-- prints its RHS at `LowestPrec`.
 type DocPrec ann = PrecedenceLevel -> Doc ann
 
+-- Specifies what kinds of operations are allowed to be printed at
+-- this point without wrapping in parens.
+data PrecedenceLevel =
+    -- Any subexpression is allowed without parens
+    LowestPrec
+    -- Function application is allowed without parens, but nothing
+    -- that binds less tightly
+  | AppPrec
+    -- Only single symbols and parens allowed
+  | ArgPrec
+  deriving (Eq, Ord)
+
 class PrettyPrec a where
   prettyPrec :: a -> DocPrec ann
 
 -- `atPrec prec doc` will ensure that the precedence level is at most
 -- `prec` when running `doc`, wrapping with parentheses if needed.
+-- To wit,
+-- - `atPrec LowestPrec` means "wrap unless the context permits all
+--   subexpressions unwrapped"
+-- - `atPrec AppPrec` means "wrap iff the context requires ArgPrec"
+-- - `atPrec ArgPrec` means "always wrap" (unless the
+--   `PrecedenceLevel` ADT is extended later).
```

Comment:
I'm not sure if I follow this. If you `atPrec ArgPrec $ Con (Lit (Int32Lit 2))` then it should be printed without parens, so "always wrap" seems incorrect?

Reply:
Good catch.  That was a typo for "never wrap".  Fixed.